### PR TITLE
Tomer/compressionable data enable

### PR DESCRIPTION
**Pull request by tomerr-lb**

### CODE REVIEW

File: protocol.cpp [hunk 1]
```diff
@@ -190,12 +192,11 @@ int redis_protocol::authenticate(const char *credentials)
...
 }
 
-int redis_protocol::write_command_set(const char *key, int key_len, const char *value, int value_len, int expiry, unsigned int offset)
+int redis_protocol::write_command_set(const char *key, int key_len, const val_list *values_list, unsigned int total_buffers_len, int expiry, unsigned int offset)
 {
     assert(key != NULL);
     assert(key_len > 0);
-    assert(value != NULL);
-    assert(value_len > 0);
+    assert(total_buffers_len > 0);
```
Comment (adir-lb) [inline protocol.cpp:199]: don't you need to assert each buffer within the list?

File: protocol.cpp [hunk 2]
```diff
@@ -240,11 +241,16 @@ int redis_protocol::write_command_set(const char *key, int key_len, const char *
             "\r\n"
             "$%u\r\n"
             "%s\r\n"
-            "$%u\r\n", (unsigned int) strlen(expiry_str), expiry_str, value_len);
+            "$%u\r\n", (unsigned int) strlen(expiry_str), expiry_str, total_buffers_len);
+    }
+
+    for (val_list_it it = values_list->begin(); it != values_list->end(); ++it)
+    {
```
Comment (adir-lb) [inline protocol.cpp]: what is the convention for one-liner for loop?
anyway i think they use to put the `{` in the same line of the `for` statement.

File: obj_gen.cpp [hunk 3]
```diff
@@ -454,9 +455,13 @@ void data_object::clear(void)
 {
     m_key = NULL;
     m_key_len = 0;
-    m_value = NULL;
-    m_value_len = 0;
+    m_total_buffers_len = 0;
     m_expiry = 0;
+    if (m_values_list)
+    {
```
Comment (sashas-lb) [inline obj_gen.cpp]:
``` c 
 if (m_values_list)  {
```

File: obj_gen.h [hunk 4]
```diff
@@ -64,8 +67,9 @@ class data_object {
     void clear(void);
     void set_key(const char* key, unsigned int key_len);
     const char* get_key(unsigned int* key_len);
-    void set_value(const char* value, unsigned int value_len);
-    const char* get_value(unsigned int* value_len);
+    void add_value(const char* value, unsigned int value_leni, bool clear_list = false);
```
Comment (adir-lb) [inline obj_gen.h]: `value_leni`?

Comment (adir-lb) [inline obj_gen.h]: `clear_list` doesn't seem related to this function.

File: protocol.h [hunk 5]
```diff
@@ -20,6 +20,10 @@
 #define _PROTOCOL_H
 
 #include <event2/buffer.h>
+#include <utility>
+#include <list>
+
+typedef std::list<std::pair<const char*, unsigned int> > val_list;
```
Comment (adir-lb) [inline protocol.h:26]: maybe your new type should contain also the total value size.

File: memtier_benchmark.cpp [hunk 6]
```diff
@@ -90,6 +90,7 @@ static void config_print(FILE *file, struct benchmark_config *cfg)
         "clients = %u\n"
         "threads = %u\n"
         "cpu_split = %s\n"
+        "compress_perc = %d\n"
```
Comment (adir-lb) [inline memtier_benchmark.cpp:93]: consider calling this param compression_ratio.

Comment (adir-lb) [reply memtier_benchmark.cpp:93]: btw, should it be `%u`?

File: obj_gen.cpp [hunk 7]
```diff
@@ -221,7 +223,14 @@ void object_generator::alloc_value_buffer(void)
...
     }
 
-    m_value_buffer_size = size;
+    if (!m_compress_perc)
+        m_value_buffer_size = size;
+    else {
+        float decimal = static_cast<float>(m_compress_perc)/100;
+        m_value_buffer_size = static_cast<unsigned int>(size*(1-decimal));
+        m_zeros_buffer_size = size - m_value_buffer_size;
+        m_zeros_buffer = (char*) calloc(m_zeros_buffer_size,1);
```
Comment (adir-lb) [inline obj_gen.cpp:305]: cosmetic: please add spaces around operators (`/`, `-`. `*`) and `,`.

File: obj_gen.cpp [hunk 8]
```diff
@@ -221,7 +223,14 @@ void object_generator::alloc_value_buffer(void)
         size = m_data_size.size_list->largest();
     }
 
-    m_value_buffer_size = size;
+    if (!m_compress_perc)
+        m_value_buffer_size = size;
+    else {
+        float decimal = static_cast<float>(m_compress_perc)/100;
```
Comment (adir-lb) [inline obj_gen.cpp:302]: consider the following - 
```cpp
float decimal = m_compress_perc / 100.0;
```

File: memtier_benchmark.cpp (hunk 6 above)
Comment (tomerr-lb) [reply memtier_benchmark.cpp:93]: yes it should
